File: hdf-engine/go/loader_test.go
package hdfengine

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
)

func loadFixtureBytes(t *testing.T, name string) []byte {
	t.Helper()
	data, err := os.ReadFile(filepath.Join("..", "testdata", name))
	if err != nil {
		t.Fatalf("read fixture %s: %v", name, err)
	}
	return data
}

func TestLoad_Results_Valid(t *testing.T) {
	data := loadFixtureBytes(t, "query-fixture.json")
	res, err := Load(data, 0)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if res.Format != FormatJSON {
		t.Errorf("format = %q, want json", res.Format)
	}
	if res.DocType != "results" {
		t.Errorf("docType = %q, want results", res.DocType)
	}
	if !res.Valid || res.Results == nil {
		t.Fatalf("expected valid results, got valid=%v results=%v err=%q", res.Valid, res.Results != nil, res.ParseError)
	}
	if res.Baseline != nil {
		t.Error("baseline should be nil for a results document")
	}
}

func TestLoad_Baseline_Valid(t *testing.T) {
	data := loadFixtureBytes(t, "baseline-fixture.json")
	res, err := Load(data, 0)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if res.DocType != "baseline" {
		t.Errorf("docType = %q, want baseline", res.DocType)
	}
	if !res.Valid || res.Baseline == nil {
		t.Fatalf("expected valid baseline, got valid=%v baseline=%v err=%q", res.Valid, res.Baseline != nil, res.ParseError)
	}
	if res.Results != nil {
		t.Error("results should be nil for a baseline document")
	}
}

func TestLoad_SizeGuardRunsFirst(t *testing.T) {
	// Oversized, non-JSON input with a tiny limit must fail with the SIZE error,
	// proving the size guard runs before any parse/detect.
	_, err := Load([]byte("this is not json and is over the tiny limit"), 4)
	if err == nil {
		t.Fatal("expected size error")
	}
	if got := err.Error(); !strings.Contains(got, "exceeds maximum") {
		t.Errorf("expected size error, got %q", got)
	}
}

func TestLoad_NDJSON_Detected(t *testing.T) {
	nd := []byte("{\"a\":1}\n{\"b\":2}\n")
	res, err := Load(nd, 0)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if res.Format != FormatNDJSON {
		t.Errorf("format = %q, want ndjson", res.Format)
	}
}

func TestLoad_PrettyPrintedSingleObject_IsJSON(t *testing.T) {
	// A pretty-printed single object spans many lines but is NOT NDJSON.
	pretty := loadFixtureBytes(t, "query-fixture.json")
	res, err := Load(pretty, 0)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if res.Format != FormatJSON {
		t.Errorf("pretty-printed object: format = %q, want json", res.Format)
	}
}

func TestLoad_InvalidResults_ReportsParseErrorNotHardFail(t *testing.T) {
	// Detected as results (has "baselines") but schema-invalid: the core returns
	// DocType + ParseError with Valid=false, and no Go error.
	bad := []byte(`{"baselines": "not an array", "components": [], "statistics": {}}`)
	res, err := Load(bad, 0)
	if err != nil {
		t.Fatalf("core should not hard-fail on invalid content: %v", err)
	}
	if res.DocType != "results" {
		t.Errorf("docType = %q, want results (detected by baselines key)", res.DocType)
	}
	if res.Valid {
		t.Error("invalid document should not be Valid")
	}
	if res.ParseError == "" {
		t.Error("expected a ParseError message for the invalid document")
	}
}

func TestLoad_UnknownType(t *testing.T) {
	res, err := Load([]byte(`{"unrecognized": true}`), 0)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if res.DocType != "" {
		t.Errorf("docType = %q, want empty for unknown type", res.DocType)
	}
	if res.Valid {
		t.Error("unknown type should not be Valid")
	}
	if res.ParseError != "" {
		t.Errorf("unknown type should not carry a ParseError, got %q", res.ParseError)
	}
}

func TestLoad_NonJSON(t *testing.T) {
	res, err := Load([]byte("not json at all"), 0)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if res.DocType != "" {
		t.Errorf("non-JSON docType = %q, want empty", res.DocType)
	}
	if res.Format != FormatJSON {
		t.Errorf("single-line non-JSON format = %q, want json (not ndjson)", res.Format)
	}
}

File: hdf-engine/test/loader.test.ts
import { describe, it, expect } from 'vitest';
import { readFileSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';
import { load, detectFormat } from '../src/loader.js';

// Shared cross-language fixtures (also read by go/loader_test.go and the Go
// parity test), so both loader cores run the same input.
const testdata = join(dirname(fileURLToPath(import.meta.url)), '..', 'testdata');
const read = (name: string): string => readFileSync(join(testdata, name), 'utf-8');

describe('load — parity with go/loader.go Load', () => {
  it('loads a valid results document', () => {
    const res = load(read('query-fixture.json'));
    expect(res.format).toBe('json');
    expect(res.docType).toBe('results');
    expect(res.valid).toBe(true);
    expect(res.results).toBeDefined();
    expect(res.baseline).toBeUndefined();
  });

  it('loads a valid baseline document', () => {
    const res = load(read('baseline-fixture.json'));
    expect(res.docType).toBe('baseline');
    expect(res.valid).toBe(true);
    expect(res.baseline).toBeDefined();
    expect(res.results).toBeUndefined();
  });

  it('runs the size guard FIRST (throws before parse)', () => {
    expect(() => load('this is not json and is over the tiny limit', 4)).toThrow(/exceeds maximum/);
  });

  it('detects NDJSON input', () => {
    expect(load('{"a":1}\n{"b":2}\n').format).toBe('ndjson');
  });

  it('a pretty-printed single object is json, not ndjson', () => {
    expect(load(read('query-fixture.json')).format).toBe('json');
  });

  it('invalid results reports parseError, not a throw', () => {
    const res = load('{"baselines": "not an array", "components": [], "statistics": {}}');
    expect(res.docType).toBe('results');
    expect(res.valid).toBe(false);
    expect(res.parseError).toBeTruthy();
  });

  it('unknown type: empty docType, not valid, no parseError', () => {
    const res = load('{"unrecognized": true}');
    expect(res.docType).toBe('');
    expect(res.valid).toBe(false);
    expect(res.parseError).toBeUndefined();
  });

  it('non-JSON: empty docType, format json (single line)', () => {
    const res = load('not json at all');
    expect(res.docType).toBe('');
    expect(res.format).toBe('json');
  });
});

// Mirror of go/loader_parity_test.go TestLoader_CrossLanguageParity: the SAME
// shared fixtures asserted to the SAME (format, docType, valid) on the TS side.
describe('cross-language parity (mirror of go/loader_parity_test.go)', () => {
  const cases: { fixture: string; format: string; docType: string; valid: boolean }[] = [
    { fixture: 'query-fixture.json', format: 'json', docType: 'results', valid: true },
    { fixture: 'baseline-fixture.json', format: 'json', docType: 'baseline', valid: true },
  ];
  for (const c of cases) {
    it(c.fixture, () => {
      const res = load(read(c.fixture));
      expect(res.format).toBe(c.format);
      expect(res.docType).toBe(c.docType);
      expect(res.valid).toBe(c.valid);
    });
  }
});

describe('detectFormat', () => {
  it('single object → json', () => {
    expect(detectFormat('{"a":1}')).toBe('json');
  });
  it('two JSON lines → ndjson', () => {
    expect(detectFormat('{"a":1}\n{"b":2}')).toBe('ndjson');
  });
  it('blank input → json', () => {
    expect(detectFormat('')).toBe('json');
  });
});
